src/ucs/sys/event_set.c

```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
     return UCS_OK;
 }
 
-ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
-                                ucs_event_set_handler_t event_set_handler,
-                                void *arg)
+unsigned ucs_event_set_wait(ucs_sys_event_set_t *event_set, int max_events,
+                            int timeout_ms,
+                            ucs_event_set_handler_t event_set_handler,
+                            void *arg)
 {
-    int nready;
+    int read_events = 0;
     struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
-    int i;
-
-    if (event_set_handler == NULL) {
-        return UCS_ERR_INVALID_PARAM;
-    }
-
-    nready = epoll_wait(event_set->epfd, ep_events, UCS_EVENT_EPOLL_MAX_EVENTS,
-                        timeout_ms);
-    if ((nready < 0) && (errno != EINTR)) {
-        ucs_error("epoll_wait() failed: %m");
-        return UCS_ERR_IO_ERROR;
-    }
-    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
-                   event_set->epfd, timeout_ms, nready);
-
-    for (i=0; i < nready; i++) {
-        int events = ucs_event_set_map_to_events(ep_events[i].events);
-        event_set_handler(ep_events[i].data.fd, events, arg);
-    }
-    return UCS_OK;
+    int i, max_wait_events, nready;
+
+    ucs_assert(event_set_handler != NULL);
+
+    do {
+        max_wait_events = ucs_min(max_events - read_events,
+                                   UCS_EVENT_EPOLL_MAX_EVENTS);
+	nready = epoll_wait(event_set->epfd, ep_events,
+                            max_wait_events, timeout_ms);
+	if (nready < 0) {
+            if (errno == EINTR) {
+                /* here we should take care about timeout_ms (i.e. decrease it),
+                 * but it is ok to not od this */
+                continue;
+            }
+//            ucs_error("epoll_wait() failed: %m");
+//	    return UCS_ERR_IO_ERROR;
+        }
+        ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
+                       event_set->epfd, timeout_ms, nready);
+
+        for (i=0; i < nready; i++) {
+            int events = ucs_event_set_map_to_events(ep_events[i].events);
+            event_set_handler(ep_events[i].data.ptr, events, arg);
+        }
+
+        read_events += nready;
+    } while ((read_events < max_events) && (nready == max_wait_events));
```

Comment:
It based on your codes. https://github.com/openucx/ucx/issues/3652#issuecomment-500110057
I used `max_events` instead of `iface->config.max_poll` 

I worried about misunderstanding your recommendation. Please let me know if I need to fix it.

Reply:
fixed